src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/AbstractConverter.cs

```diff
@@ -93,5 +114,166 @@ private static QueryClauseSyntax CreateQueryClause(ExtendedSyntaxNode node)
+                //
+                // INPUT:
+                //   foreach (var n1 in c1)
+                //      foreach (var n2 in c2)
+                //          ...
+                //
+                // OUTPUT:
+                //   c1.SelectMany(n1 => c2.Select(n2 => ...
+                //
+                case SyntaxKind.ForEachStatement:
+                    hasForEachChild = true;
+                    var foreachStatement = (ForEachStatementSyntax)node.Node;
+                    ++extendedNodeIndex;
+                    return CreateLinqInvocation(
+                        foreachStatement,
+                        receiverForInvocation: foreachStatement.Expression,
+                        selectExpression: selectExpression,
+                        leadingCommentsTrivia: node.ExtraLeadingComments,
+                        trailingCommentsTrivia: node.ExtraTrailingComments,
+                        currentExtendedNodeIndex: ref extendedNodeIndex);
+
+                // Nested If statement is converted into a Where method invocation on the current receiver. For example:
+                //
+                // INPUT:
+                //   foreach (var n1 in c1)
+                //      if (n1 > 0)
+                //          ...
+                //
+                // OUTPUT:
+                //   c1.Where(n1 => n1 > 0).Select(n1 => ...
+                //
+                case SyntaxKind.IfStatement:
+                    var ifStatement = (IfStatementSyntax)node.Node;
+                    var parentForEachStatement = ifStatement.GetAncestor<ForEachStatementSyntax>();
+                    var lambdaParameter = SyntaxFactory.Parameter(SyntaxFactory.Identifier(parentForEachStatement.Identifier.ValueText));
+                    var lambda = SyntaxFactory.SimpleLambdaExpression(
+                        SyntaxFactory.Parameter(
+                            SyntaxFactory.Identifier(parentForEachStatement.Identifier.ValueText)),
+                        ifStatement.Condition.WithCommentsFrom(ifStatement.OpenParenToken, ifStatement.CloseParenToken))
+                        .WithCommentsFrom(ifStatement.IfKeyword.GetAllTrivia().Concat(node.ExtraLeadingComments), node.ExtraTrailingComments);
+
+                    receiver = SyntaxFactory.InvocationExpression(
+                        SyntaxFactory.MemberAccessExpression(
+                            SyntaxKind.SimpleMemberAccessExpression,
+                            receiver.Parenthesize(),
+                            SyntaxFactory.IdentifierName(nameof(Enumerable.Where))),
+                        SyntaxFactory.ArgumentList(SyntaxFactory.SingletonSeparatedList(
+                            SyntaxFactory.Argument(lambda))));
+
+                    ++extendedNodeIndex;
+                    return CreateLinqInvocationForExtendedNode(selectExpression, ref extendedNodeIndex, ref receiver, ref hasForEachChild);
+            }
+
+            throw ExceptionUtilities.Unreachable;
```

Comment:
this worries me a bit.  can we doc how we know this is really unreachable?

Reply:
Sure. This is guaranteed by the existing code that builds up this converter. Ivan's existing query form refactoring guarantees it [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/ConvertLinq/ConvertForEachToLinqQuery/AbstractConverter.cs,47), and the newly added refactoring bails out in presence of local declarations, so the only possible cases are nested foreach and if statements. I will add more comments to clarify.